subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java

```diff
@@ -177,11 +235,34 @@ public SnapshotVisitResult visitEntry(FileSystemLocationSnapshot snapshot) {
 
         @Override
         public void leaveDirectory(DirectorySnapshot directorySnapshot) {
-            visitor.postDirectory();
+            DirectorySnapshot lastUnvisitedDirectory = unvisitedDirectories.pollLast();
+            if (lastUnvisitedDirectory == null) {
+                visitor.postDirectory();
+            }
+
             if (--depth == 0) {
                 visitor.postRoot();
             }
         }
+
+        private void visitUnvisitedDirectories() {
+            DirectorySnapshot unvisited = unvisitedDirectories.poll();
+            while (unvisited != null) {
```

Comment:
You could use a for loop here and then call `unvisitedDirectories.clear()` afterwards.

Reply:
Fine with me.